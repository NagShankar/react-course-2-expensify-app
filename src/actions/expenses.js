import uuid from 'uuid4';


//ADD_EXPENSE action creator
export const addExpense = ({description='', note='', amount=0, createdAt=0}={}) => {
    return {
        type:"ADD_EXPENSE",
        expense:{
            id:uuid(),
            description, //or description= description, or use shortand property
            note,
            amount,
            createdAt
        }
        
    }
}


//REMOVE_EXPENSE
export const removeExpense = ({id}={}) => {
    return {
        type:"REMOVE_EXPENSE",
        id //or id:id
        
    }
}



//EDIT_EXPENSE
export const editExpense = (id, updates) => {
    return {
        type:"EDIT_EXPENSE",
        id,
        updates
        
    }
}


//we are doing named export