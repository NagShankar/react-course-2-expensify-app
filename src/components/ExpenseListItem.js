import React from "react";
//import { connect } from 'react-redux';
//import { removeExpense } from './../actions/expenses'
import { Link } from 'react-router-dom';

//now destructuring instead od taking props
const ExpenseListItem = ({id, description, amount, createdAt}) => {
  return(
    <div>
          <Link to={`/edit/${id}`}><h3>{description} </h3></Link>
          <h4>Amount: {amount} Created: {createdAt}</h4>
         
          {/* lets have remove button only when wants to edit an expense*/} 
         
          {/* 
              <button onClick={()=>{
                    dispatch(removeExpense({id:id})); //remove expense accepts "id" as an object, check actions/expenses.js for more details, using dispatch after destructuring which was available inside props as props.dispatch
               }}>Remove</button>

           */} 
      </div>
  );    
}

//export default connect()(ExpenseListItem); //CONNECTING GIVE DISPATCH AS A PROP

export default ExpenseListItem;