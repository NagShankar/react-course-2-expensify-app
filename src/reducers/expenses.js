//no dependecy here to import anything because reducers are PURE functions


//Expense Reducer
const expensesReducerDefaultState = [];

const expensesReducer = (state=expensesReducerDefaultState, action) => {
  switch(action.type){
      case 'ADD_EXPENSE' :
          //return state.concat(action.expense), lets use SPEARD syntax like below
          return [
              ...state,
              action.expense
          ]
      case 'REMOVE_EXPENSE':
          return state.filter((expense)=>{
                  return expense.id !== action.id
              })
          
      case 'EDIT_EXPENSE':
          return state.map((expense)=>{
                  if(expense.id === action.id){
                    return {
                        ...expense,
                        ...action.updates
                    }                     
                   }else{
                      return expense
                  }
              })
                 
      default:
          return state;
  }  
};


export default expensesReducer;