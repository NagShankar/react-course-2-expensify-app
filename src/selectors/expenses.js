import moment from 'moment';

const getVisibleFilters = ((expenses, {text, sortBy, startDate, endDate}) => {
    
    
/*  

In your filters, startDate and endDate can be undefined. In the case that the startDate or endDate for your filters is undefined, you do not want to filter out any data based on that criteria. If startDate is undefined, typeof startDate !== 'number' will be true, and startDateMatch will be true.

In the case that you provide a number to startDateMatch, that means that you do want to filter by date. That means that the first part of your or (||) condition will be false and you will be checking for a match with the second part of the condition (whether the expense date comes before or after the start/endDate).

In short, undefined start and end dates will have typeof date !== 'number' be true and let all data pass through the filter (assuming the expense meets the other criteria as well!!). Start dates and end dates with values will have this condition be false, and you will filter the expenses based on createdAt.

*/
    
    
    return expenses.filter((expense)=>{
        
//        const startDateMatch = typeof startDate !== 'number' || expense.createdAt >= startDate;//before comparison was done using numbers, now we need to use actual date
//        const endDateMatch = typeof endDate !== 'number' || expense.createdAt <= endDate;//before comparison was done using numbers, now we need to use actual date
        const createdAtMoment = moment(expense.createdAt);
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;//first checking if start date even exist because we have given user to clear range by clicking "X", else setting true to pass the final triple && condition
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;//first checking if start date even exist because we have given user to clear range by clicking "X", else setting true to pass the final triple && condition
        const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());
                       
        return startDateMatch && endDateMatch && textMatch
    }).sort((a,b) => {
      if(sortBy==="date"){
          return a.createdAt < b.createdAt ? 1 : -1
      }
        else if(sortBy==="amount"){
          return a.amount < b.amount ? 1 : -1
      }
        
    })
});

export default getVisibleFilters;