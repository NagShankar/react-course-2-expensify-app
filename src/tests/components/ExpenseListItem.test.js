import React from "react";
import { shallow } from 'enzyme';
import ExpenseListItem  from '../../components/ExpenseListItem'; //there's no connected vresion, hence importing default
import expenses from '../fixtures/expenses'; //our own data for testing puspose

test('should render single list item',()=>{
    const wrapper = shallow(<ExpenseListItem id={expenses[2].id} description={expenses[2].description} amount={expenses[2].amount} createdAt={expenses[2].createdAt}/>);
    expect(wrapper).toMatchSnapshot();
})

//or simply spread an expense like below
//test('should render single list item',()=>{
//    const wrapper = shallow(<ExpenseListItem {...expenses[0]}/>);
//    expect(wrapper).toMatchSnapshot();
//})