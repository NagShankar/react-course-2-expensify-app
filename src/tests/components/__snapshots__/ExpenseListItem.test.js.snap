// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render single list item 1`] = `
<div>
  <Link
    to="/edit/3"
  >
    <h3>
      Credit Card
       
    </h3>
  </Link>
  <h4>
    Amount: 
    4500
     Created: 
    345600000
  </h4>
</div>
`;
